proxy/src/connection.rs

```diff
@@ -30,6 +30,18 @@ pub fn connect(addr: &SocketAddr,
     }
 }
 
+/// A server socket that is in the process of conditionally upgrading to TLS.
+enum ConditionallyUpgradeServerToTls {
```

Comment:
I notice that this is essentially doing the same thing for accepted connections that `Connecting` does for connect, is there common functionality that could (eventually) be factored out?

Reply:
(Also, I find this name somewhat too verbose, but that could just be a matter of taste...)

Reply:
> I notice that this is essentially doing the same thing for accepted connections that Connecting does for connect, is there common functionality that could (eventually) be factored out?

I think we could factor out some common logic. It is a little trickier than it might seem at first, because `connect()` needs the TLS identity at a different time than `accept()` does, and because `accept()` does the protocol discrimination and `connect()` doesn't. So abstracting might be a win or it might just make things more complicated.

Reply:
Yeah, that's fair --- just something to think about in the future.